### ISSUE TITLE
Hyphen and en dash

### ISSUE THREAD

**Issue by tflo** [source-foundry/Hack#426]
Something seems wrong with the hyphen and the en dash:

This is Hack on macOS 10.13.4:

![screen shot 2018-04-12 at 01 29 23](https://user-images.githubusercontent.com/1410282/38648428-6cfa4634-3df1-11e8-9038-2d3002464e28.png)

IMO the en dash should be visibly larger than the hyphen, not reverse.

It’s the Hack release version (Version 3.003;[3114f1256]-release; ttfautohint (v1.7) -l 6 -r 50 -G 200 -x 10 -H 181 -D latn -f latn -m "Hack-Regular-TA.txt" -w G -W -t -X "")

For comparison, this is DejaVu Sans Mono, which looks more reasonable (hyphen and en dash):

![screen shot 2018-04-12 at 01 30 06](https://user-images.githubusercontent.com/1410282/38648560-3541317a-3df2-11e8-9fa5-aa9c70f84ce6.png)

– Tom

**Comment by jdw1996** [comment]
I agree that the en dash should probably be longer than the hyphen, but I also don't like how short the hyphen is in DejaVu Sans Mono.  I'd be OK with the hyphen being a bit shorter than it is at the moment, but I can't think of a situation where I'm looking at an en dash (or an em dash) in a monospaced font.  Out of curiosity, what is the use case where you noticed this?

**Comment by chrissimpkins** [comment]
This was addressed a few versions ago.  Here is what we landed on: https://github.com/source-foundry/Hack/issues/178#issuecomment-350862403

Happy to re-evaluate if you feel that it needs to be modified further.

**Comment by chrissimpkins** [comment]
And note that others have had complaints that some of these look too similar... https://github.com/source-foundry/Hack/issues/178#issuecomment-247572689 

I suppose it depends upon use cases.  The goal at that time was to distinguish the different code points.

**Comment by tflo** [comment]
> This was addressed a few versions ago. Here is what we landed on: #178 (comment)

I’ve seen that discussion, but something must have changed, because in [the screenshot](https://github.com/source-foundry/Hack/issues/178#issuecomment-247572689) posted by @jwheare the hyphen is clearly shorter than the en dash. (The rest of the discussion is mainly about the length of the figure dash, a char that I’m not using.)

I’m aware of the benefit of a better visible, longish hyphen character, since in code you are not using any other dashes and thus making it look like an en dash seems like a good solution.

My situation is that I’m using Hack also for TeX (ConTeXt) with prose in German. In German both the en dash and the hyphen is used frequently (the en dash has —roughly— the same role as the em dash in US English), and so the inverted dimensions of the two chars are confusing.

On the other hand the same ConTeXt source usually also contains code (TeX and Lua).

So, my ideal solution goes in this direction:

- in any case the en dash must not be shorter than the hyphen
- hyphen should be (slightly) better visible than in DejaVu Mono

This screenshot is IBM Plex Mono, which seems like a not too bad solution to me:

![screen shot 2018-04-13 at 14 02 11](https://user-images.githubusercontent.com/1410282/38733862-da94f198-3f23-11e8-995f-edd270333dce.png)

– Tom

**Comment by jdw1996** [comment]
In LaTeX, at least, you can use two hyphens (`--`) to get an en dash (and three for an em dash).  I understand if that doesn't work in the implementation of TeX you're using, or if you just prefer using Unicode characters, but otherwise it might help reduce ambiguity.

**Comment by tflo** [comment]
Well, to sum it up, what I think is this:

- It is an absolute no-go to to make the hyphen (U+002D) longer than the en dash (U+2013), even for a monospaced font.
- It is important (since the main purpose of the font is code) to have a good visibility of the hyphen; DejaVu’s hyphen is pretty tiny, so, it must be something better visible, but _without_ gulping the en dash; or even inverting the dimensions.

– Tom

**Comment by l3u** [comment]
I recently started to use Hack and this issue hit me – I first thought, the way to input an en dash has been changed, but then I found this discussion.

The issue is at least still presend in version 3.003:

![hypen-ndash](https://user-images.githubusercontent.com/6770781/52262561-13904580-292d-11e9-8d8a-bb511a293ed0.png)

Clearly, the Hyphen-Minus is longer than the En Dash – and it really should be vice versa. Would be nice if this could be fixed … it almost stops me from using Hack at all …

**Comment by chrissimpkins** [comment]
While we are at it, can we discuss vertical orientation for the en/em dashes too? May as well address that if changes are needed while we are working on these shapes!

**Comment by l3u** [comment]
Here's the clear text of the (KWrite) screenshot I posted above:
```
U+002D - Hyphen-Minus
U+2013 – En Dash
```
Surely, the en dash is not commonly used in source code, but I e. g. noticed this whilst working on a (German) rst file source where I use it. And also inside text-only emails.

I don't know what's typographically correct, but I think LaTeX is always correct ;-) So here's what LaTeX does:
![dashes](https://user-images.githubusercontent.com/6770781/52295762-e3709300-297c-11e9-9f0d-3a0c8fe6260f.png)
The hyphen is a bit thicker and a bit below the en and the em dash, which are a bit thinner. So speaking of your image: the hyphen should be as wide as the en dash is now, the en dash should be as wide as the hyphen is now, the en and em dashes should be as high as the hyphen is now and the hyphen should be as low as the en and em dashes are now. Here we are ;-)

**Comment by chrissimpkins** [comment]
I built a set of fonts with changes in the **regular set only** that get us here (displayed as em - en - hyphen):

![h8j18-image](https://user-images.githubusercontent.com/4249591/52389061-55b9a400-2a5f-11e9-9472-c329ad0a3a99.png)

![3d3iv-image](https://user-images.githubusercontent.com/4249591/52389079-6c5ffb00-2a5f-11e9-858a-cdbf454a214d.png)

Attached as zip archive below and based on the v4.000 changes in https://github.com/source-foundry/Hack/pull/457 so you will see other shape changes that have happened since v3.003.  Let me know what you think.  

[ttf.zip](https://github.com/source-foundry/Hack/files/2839215/ttf.zip)

**Comment by l3u** [comment]
This looks way better :-) I would still swap the vertical orientation (as LaTeX does) though.

**Comment by chrissimpkins** [comment]
> I would still swap the vertical orientation (as LaTeX does) though.

Will give this a try, make the mods across sets and push a new set of fonts that you can use for now.  These changes will make their way into the next release.  Thanks again for the feedback here!

**Comment by l3u** [comment]
Nice :-) Thanks a lot for fixing this!

**Comment by tflo** [comment]
Luckily I came across this old thread while cleaning up my issue subscriptions. In the meantime I had abandoned Hack, mainly because of the faulty en dash.

> Attached as zip archive below and based on the v4.000 changes in #457 so you will see other shape changes that have happened since v3.003. Let me know what you think.

This… is very nice and virtually perfect!

***

For comparison again the old and the new version:

![Hack-old](https://user-images.githubusercontent.com/1410282/60332867-f6214e00-9997-11e9-9d4a-995d09112684.png)

![Hack-new](https://user-images.githubusercontent.com/1410282/60332901-05a09700-9998-11e9-93aa-98328c26a9e1.png)

***

And here some real-world prose:

_Confusing:_
![Screen Shot 2019-06-28 at 11 16 04-pty](https://user-images.githubusercontent.com/1410282/60332937-19e49400-9998-11e9-8bde-fa69375ef0c8.png)

_This works:_
![Screen Shot 2019-06-28 at 11 17 53-pty](https://user-images.githubusercontent.com/1410282/60333022-5617f480-9998-11e9-87db-7a3595c2aeb1.png)

***

IMO, the new solution is quite perfect in all criteria:

- En dash not shorter than hyphen (important for monospaced prose, for example when doing TeX)
- Hyphen has has better visibility than with Deja Vu (important for normal code)
- Bonus: em dash clearly distinguishable from en dash

I think, it’s even better than IBM Plex now!

Many thanks for the update!

– Tom
